### PR TITLE
Add templating, via jinja2, to cephadm

### PR DESCRIPTION
**Pull request by phlogistonjohn**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), <s>opened tracker ticket</s>
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by phlogistonjohn** [comment]
Open question:  If we are going to add external deps to cephadm and will do so via pip, should we also include hashes in our requreiments file? Example:
```
jinja2==3.1.2 \
    --hash=sha256:31351a702a408a9e7595a8fc6150fc3f43bb6bf7e319770cbc0db9df9437e852
markupsafe==2.1.3 \
    --hash=sha256:af598ed32d6ae86f1b747b82783958b1a4ab8f617b06fe68795c7f026abbdcad
```

File: src/cephadm/cephadm.py [hunk 1]
```diff
@@ -3234,58 +3235,68 @@ def install_base_units(ctx, fsid):
...
-                      'PIDFile=%t/%n-pid\n')
-        if ctx.container_engine.version >= CGROUPS_SPLIT_PODMAN_VERSION:
-            extra_args += 'Delegate=yes\n'
-
-    docker = isinstance(ctx.container_engine, Docker)
-    u = """# generated by cephadm
+_TMPL_DAEMON_UNIT = """# generated by cephadm
 [Unit]
-Description=Ceph %i for {fsid}
+Description=Ceph %i for {{fsid}}
```
Comment (mgfritch) [inline src/cephadm/cephadm.py:3240]: maybe now is a good time to formally move these strs into a `.j2` (or `.jinja`) template file?

Comment (phlogistonjohn) [reply src/cephadm/cephadm.py:3240]: I think that's doable. I was looking at the jinj2 docs (and source) recently and the package loader appears to support zipimport (which IIUC is the basis for zipapp).

What I'll do is try this out in a an experimental branch and if all goes smoothly I'll fold that branch back into this one. If things don't work out I'll report my findings here.

**Comment by phlogistonjohn** [comment]
After having a discussion with our build team I'm happy to be able to more forward with this PR. This PR is now part 1 of 2. This PR adds the basics of bundled dependencies in cephadm and anyone familiar with python should be able to follow along.

The follow up PR will add an additional source for the bundled dependencies, and the ability to turn them off entirely. It also plumbs that through cmake and into ceph's rpm spec.in file, as well as enhancing the `cephadm version` command with a flag to display more information about the contents of the zipapp.

**Comment by phlogistonjohn** [comment]
I think I have resolved the CI issues related to building on systems without pip. Now, either pip or venv is required, but if both are missing things will break. I chose to handle this inside the build.py script because that's in python and the alternative was handling it in cmake and I find working with cmake far less pleasant. It also makes build.py somewhat self-contained.

@mgfritch PTA(nother)L

@adk3798 I think this is ready for teuthology testing at your earliest convenience. PTAL as well.

You may also want to peek at #54173, for next steps, but for now I think this is able to stand on it's own in a purely ceph-upstream context.

**Review by adk3798 — APPROVED** [review]
Seems good to me. Biggest concern is always bringing in the deps itself, but we've already discussed that at length.

**Comment by adk3798** [comment]
Build seems to have failed however
```
cephadm/build.py: Installing dependencies
Collecting MarkupSafe==2.1.3 (from -r zipapp-reqs.txt (line 10))
  Could not find a version that satisfies the requirement MarkupSafe==2.1.3 (from -r zipapp-reqs.txt (line 10)) (from versions: 0.9, 0.9.1, 0.9.2, 0.9.3, 0.11, 0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.20, 0.21, 0.22, 0.23, 1.0, 1.1.0, 1.1.1, 2.0.0a1, 2.0.0rc1, 2.0.0rc2, 2.0.0, 2.0.1)
No matching distribution found for MarkupSafe==2.1.3 (from -r zipapp-reqs.txt (line 10))
Traceback (most recent call last):
  File "build.py", line 236, in <module>
    main()
  File "build.py", line 232, in main
    _build(dest, source, versioning_vars=args.version_vars)
  File "build.py", line 65, in _build
    _install_deps(tempdir)
  File "build.py", line 155, in _install_deps
    env=env,
  File "/usr/lib64/python3.6/subprocess.py", line 311, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3.6', '-m', 'pip', 'install', '--no-binary', ':all:', '--requirement', 'zipapp-reqs.txt', '--target', PosixPath('/tmp/tmpi4ns42j4.cephadm.build')]' returned non-zero exit status 1.
make[2]: *** [src/cephadm/CMakeFiles/cephadm.dir/build.make:74: bin/cephadm] Error 1
make[1]: *** [CMakeFiles/Makefile2:7303: src/cephadm/CMakeFiles/cephadm.dir/all] Error 2
```
centos 8
https://shaman.ceph.com/builds/ceph/wip-adk-testing-2023-10-30-1319/fef23a2cd9927e4cebe919f1a4b1d998d9c1cff5/default/364525/

centos 9
https://shaman.ceph.com/builds/ceph/wip-adk-testing-2023-10-30-1319/fef23a2cd9927e4cebe919f1a4b1d998d9c1cff5/default/364555/

ubuntu focal
https://shaman.ceph.com/builds/ceph/wip-adk-testing-2023-10-30-1319/fef23a2cd9927e4cebe919f1a4b1d998d9c1cff5/default/364563/

**Comment by phlogistonjohn** [comment]
jenkins test dashboard cephadm

File: src/cephadm/cephadmlib/templates/agent.service.j2 [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# generated by cephadm
+[Unit]
+Description=cephadm agent for cluster {{agent.fsid}}
+
+PartOf=ceph-{{agent.fsid}}.target
+Before=ceph-{{agent.fsid}}.target
+
+[Service]
+Type=forking
+ExecStart=/bin/bash {{agent.data_dir}}/unit.run
```
Comment (adk3798) [inline src/cephadm/cephadmlib/templates/agent.service.j2:10]: Seems to be an issue with the agent running off of this. 
```
Oct 31 13:59:17 smithi092 systemd[1]: Starting cephadm agent for cluster e3a2336e-77ed-11ee-8db9-212e2dc638e7...
Oct 31 13:59:17 smithi092 bash[146780]: /bin/bash: /unit.run: No such file or directory
Oct 31 13:59:17 smithi092 systemd[1]: ceph-e3a2336e-77ed-11ee-8db9-212e2dc638e7@agent.smithi092.service: Control process exited, code=exited status=127
Oct 31 13:59:17 smithi092 systemd[1]: ceph-e3a2336e-77ed-11ee-8db9-212e2dc638e7@agent.smithi092.service: Failed with result 'exit-code'.
Oct 31 13:59:17 smithi092 systemd[1]: Failed to start cephadm agent for cluster e3a2336e-77ed-11ee-8db9-212e2dc638e7.
Oct 31 13:59:27 smithi092 systemd[1]: ceph-e3a2336e-77ed-11ee-8db9-212e2dc638e7@agent.smithi092.service: Service RestartSec=10s expired, scheduling restart.
Oct 31 13:59:27 smithi092 systemd[1]: ceph-e3a2336e-77ed-11ee-8db9-212e2dc638e7@agent.smithi092.service: Scheduled restart job, restart counter is at 4.
Oct 31 13:59:27 smithi092 systemd[1]: Stopped cephadm agent for cluster e3a2336e-77ed-11ee-8db9-212e2dc638e7.
Oct 31 13:59:27 smithi092 systemd[1]: Starting cephadm agent for cluster e3a2336e-77ed-11ee-8db9-212e2dc638e7...
Oct 31 13:59:27 smithi092 bash[146785]: /bin/bash: /unit.run: No such file or directory
Oct 31 13:59:27 smithi092 systemd[1]: ceph-e3a2336e-77ed-11ee-8db9-212e2dc638e7@agent.smithi092.service: Control process exited, code=exited status=127
Oct 31 13:59:27 smithi092 systemd[1]: ceph-e3a2336e-77ed-11ee-8db9-212e2dc638e7@agent.smithi092.service: Failed with result 'exit-code'.
Oct 31 13:59:27 smithi092 systemd[1]: Failed to start cephadm agent for cluster e3a2336e-77ed-11ee-8db9-212e2dc638e7.
```

```
[root@smithi092 ~]# cat /etc/systemd/system/ceph-e3a2336e-77ed-11ee-8db9-212e2dc638e7@agent.smithi092.service 
# generated by cephadm
[Unit]
Description=cephadm agent for cluster e3a2336e-77ed-11ee-8db9-212e2dc638e7

PartOf=ceph-e3a2336e-77ed-11ee-8db9-212e2dc638e7.target
Before=ceph-e3a2336e-77ed-11ee-8db9-212e2dc638e7.target

[Service]
Type=forking
ExecStart=/bin/bash /unit.run
Restart=on-failure
RestartSec=10s

[Install]
WantedBy=ceph-e3a2336e-77ed-11ee-8db9-212e2dc638e7.target
```
https://pulpito.ceph.com/adking-2023-10-31_04:11:54-orch:cephadm-wip-adk-testing-2023-10-30-1928-distro-default-smithi/7441777/

**Review by adk3798 — COMMENTED** [review]
Builds are now all working. But seeing some test failures.

File: src/cephadm/zipapp-reqs.txt [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+#
+# IMPORTANT: The cephadm binary is expected to be portable across python
+# versions and CPU architectures. Dependencies are copied into the zipapp
+# by the build script and must not require compiled C (or C++, Rust, etc)
+# modules. Modules that have an optional C accelerator but can fall back
+# to pure python are OK. When you add a package to this list verify that
+# build.py creates the zipapp with only python files.
+#
+MarkupSafe >= 2.0.1 ; python_version < "3.7"
+MarkupSafe >= 2.1.3, <2.2 ; python_version >= "3.7"
```
[the remark this answers is not in the document]
Comment (phlogistonjohn) [reply src/cephadm/zipapp-reqs.txt:11]: There's a bug(?) in some versions of pip that cause it to fail to install if you disable binary packages (wheels), as it just tries to use pkg a during the install of pkg b w/o actually installing a first.
This was occurring on ubuntu 20.04 but I worked around it by upgrading pip *if* we were using a virtualenv. Unfortunately, build.py seems to be quickly accumulating a lot of hacks, largely from having to deal with the unfortunate variety of envs ceph has in it's build and test.

Do you think I should try and switch the behavior from using a virtualenv only if pip is unavailable to using a virtualenv if `venv` is available and only falling back to non-venv pip?

**Comment by adk3798** [comment]
https://pulpito.ceph.com/phlogistonjohn-2023-11-02_14:19:20-orch:cephadm-wip-phlogistonjohn-testing-2023-11-01-2015-distro-default-smithi/

reruns of failed jobs: https://pulpito.ceph.com/adking-2023-11-05_15:58:41-orch:cephadm-wip-phlogistonjohn-testing-2023-11-01-2015-distro-default-smithi/

After reruns, 1 failure:

- mgr-nfs-upgrade test, known failure for a while now, typically due to https://tracker.ceph.com/issues/62958 although this time it seems to have timed out instead of failing immediately

File: src/cephadm/cephadmlib/templates/ceph.service.j2
Comment (mgfritch) [inline src/cephadm/cephadmlib/templates/ceph.service.j2:1]: I would imagine at some point a user might want to override this file similar to the [service templates](https://docs.ceph.com/en/latest/cephadm/services/monitoring/#option-names) in the mgr

**Review by mgfritch — APPROVED** [review]

**Pull request merged**